File: Palindrome_Partitioning.py
'''
Given a string s, partition s such that every substring of the partition is a palindrome. Return all possible palindrome partitioning of s.

A palindrome string is a string that reads the same backward as forward.



Example 1:

Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]
Example 2:

Input: s = "a"
Output: [["a"]]


Constraints:

1 <= s.length <= 16
s contains only lowercase English letters.
'''

'''
Solution 2: DP
use cache to store if i,j is palindrome

dp[n] is [[String]]() type object
dp[i+1]:
- if left,i is palindrome, for loop dp[left], append s[left...i] to dp[i+1]

Time Complexity:
O(n* 2^n)

Space Complexity:
O(n^2)
'''
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        dp = [[] for _ in range(n+1)]
        # cache[i][j] = true if ispalindrome
        cache = [[False for _ in range(n)] for _ in range(n)]
        for i in range(n):
            for l in range(i+1):
                if s[l] == s[i] and (i-l<=1 or cache[l+1][i-1] == True):
                    cache[l][i] = True
                    sub = s[l:i+1:]
                    if len(dp[l]) == 0:
                        dp[i+1].append([sub])
                        continue
                    for pre in dp[l]:
                        temp = pre.copy() # copy to avoid reference miss
                        temp.append(sub)
                        dp[i+1].append(temp)

        return dp[n]

File: Palindrome_Partitioning.java
/*131. Palindrome Partitioning   Add to List QuestionEditorial Solution  My Submissions
Total Accepted: 85483
Total Submissions: 275022
Difficulty: Medium
Contributors: Admin
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

For example, given s = "aab",
Return

[
  ["aa","b"],
  ["a","a","b"]
]
Hide Company Tags Bloomberg
Hide Tags Backtracking
Hide Similar Problems (H) Palindrome Partitioning II
*/




/////////////////////////////////////////////////////////////////////////////////////
//C++




/*
Solution 1 is faster

Solution 1: Backtracking and DFS
1. backtrack the string, start from 0
2. check if the last string is palindrome string
3. if it is, check until the end, and push them into the currList
4. after check the last string, remove the last elements

Solution 2: DP
use a ret[] for List<List<String>>
then at the end return the ret[len], len = s.length()
ret[i] store from the begining until current index i, all possible partitions
use a pair to check if the substring is a pal, if pair[i][j] is true,
means the substring from i to j is pal, which means i to j is palindrome.
 */
public class Solution {
    private List<List<String>> retList = new ArrayList<List<String>>();
    private List<String> curList = new ArrayList<String>();

    public List<List<String>> partition(String s) {
        backTrack(s, 0);
        return retList;
    }

    public void backTrack(String s, int index){
        if(curList.size()>0 && index>=s.length()) { //have already get one solution, add it to return list
            List<String> temp = new ArrayList<>(curList);
            retList.add(temp);
        }

        for(int i = index; i < s.length(); ++i){
            if(isPalindrome(s, index, i)){
                String str = s.substring(index, i + 1);
                curList.add(str);
                backTrack(s, i + 1);
                /*System.out.println("curList: ");
                for(String s1 : curList){
                    System.out.println(s1 + " ");
                }*/
                curList.remove(curList.size()-1); //remember to remove the last element
            }
        }
    }

    //check if (index, i) is palindrome in string s
    public boolean isPalindrome(String s, int s1, int e1){
        for(int a = s1, b = e1; a < b; a++, b-- ){
            if(s.charAt(a) != s.charAt(b)){
                return false;
            }
        }
        return true;
    }
}

//Solution 2
public class Solution {
    public List<List<String>> partition(String s) {
        if(s.length() <= 0) return new ArrayList<>();

        int len = s.length();
        List<List<String>>[] ret = new List[len+1];
        ret[0] = new ArrayList<List<String>>();
        ret[0].add(new ArrayList<String>());

        boolean[][] pair = new boolean[len][len]; //help check if the former string is palindrome
        for(int i = 0; i < len; ++i){
            ret[i+1] = new ArrayList<List<String>>();
            for(int left = 0; left <= i; ++left){
                if(s.charAt(left) == s.charAt(i) && (i-left<=1 || pair[left+1][i-1]==true)){ //remember left+1, i-1, donot change the plus or minus
                    pair[left][i] = true;
                    String str = s. substring(left, i + 1); //get the current palindrome string
                    for(List<String> r : ret[left]){
                        List<String> temp = new ArrayList<String>(r); //construct a list contains the element in the r
                        temp.add(str);
                        ret[i+1].add(temp);
                    }
                }
            }
        }

        return ret[len];
    }
}

